lib/spack/llnl/util/symlink.py

```diff
@@ -2,77 +2,174 @@
+def symlink(source_path: str, link_path: str):
     """
-    Create a symbolic link.
+    Create a link.
+
+    On non-Windows and Windows with System Administrator
+    privleges this will be a normal symbolic link via
+    os.symlink.
+
+    On Windows without privledges the link will be a
+    junction for a directory and a hardlink for a file.
+    On Windows the various link types are:
+
+    Symbolic Link: A link to a file or directory on the
+    same or different volume (drive letter) or even to
+    a remote file or directory (using UNC in its path).
+    Need System Administrator privileges to make these.
 
-    On Windows, use junctions if os.symlink fails.
+    Hard Link: A link to a file on the same volume (drive
+    letter) only. Every file (file's data) has at least 1
+    hard link (file's name). But when this method creates
+    a new hard link there will be 2. Deleting all hard
+    links effectively deletes the file. Don't need System
+    Administrator privileges.
+
+    Junction: A link to a directory on the same or different
+    volume (drive letter) but not to a remote directory. Don't
+    need System Administrator privileges.
     """
-    if sys.platform != "win32":
-        os.symlink(real_path, link_path)
-    elif _win32_can_symlink():
-        # Windows requires target_is_directory=True when the target is a dir.
-        os.symlink(real_path, link_path, target_is_directory=os.path.isdir(real_path))
-    else:
-        try:
-            # Try to use junctions
-            _win32_junction(real_path, link_path)
-        except OSError as e:
-            if e.errno == errno.EEXIST:
-                # EEXIST error indicates that file we're trying to "link"
-                # is already present, don't bother trying to copy which will also fail
-                # just raise
-                raise
+    source_path = os.path.normpath(source_path)
+    win_source_path = source_path
+    link_path = os.path.normpath(link_path)
+
+    # Perform basic checks to make sure symlinking will succeed
+    if os.path.lexists(link_path):
+        raise SymlinkError(f"Link path ({link_path}) already exists. Cannot create link.")
+
+    if not os.path.exists(source_path):
+        if os.path.isabs(source_path):
+            # An absolute source path that does not exist will result in a broken link.
+            raise SymlinkError(
+                f"Source path ({source_path}) is absolute but does not exist. Resulting "
+                f"link would be broken so not making link."
+            )
+        else:
+            # os.symlink can create a link when the given source path is relative to
+            # the link path. Emulate this behavior and check to see if the source exists
+            # relative to the link patg ahead of link creation to prevent broken
+            # links from being made.
+            link_parent_dir = os.path.dirname(link_path)
+            relative_path = os.path.join(link_parent_dir, source_path)
```

Comment:
if `link_path` is absolute:

> `link_path = '/path/to/source/file'`
> `relative_path = '/path/to/source/file/../path/to/link'`

if `link_path` is relative:
> `link_path = 'to/source/file'`
> `relative_path = 'to/source/file/../path/to/link'`

either way, `relative_path` is relative to the location of the source file. But happy to change the variable name if you have a suggestion.

This is the same way os.symlink makes links with relative targets, so no, it doesn't fail.

Reply:
Lets say we have a dir structure like

```
root/
  x1/
    y1/
      z1
```

and we call `symlink(x1/y1/z1, s1)` from `root/`, in this case `link_parent_dir` is `.`, `relative_path == './x1/y1/z1'`, and the symlink will ultimately point to `./x1/y1/z1`. On Linux this would be considered broken (since the link target is resolved relative to link) and I assume it would be broken on Windows as well (I assume the link targets are not resolved relative to our CWD).

Reply:
> relative_path == './x1/y1/z1', and the symlink will ultimately point to ./x1/y1/z1

Note that this only checks that `./x1/y1/z1` exists from the CWD, it doesn't set the link target (`source_path`) on linux or Mac.

> On Linux this would be considered broken (since the link target is resolved relative to link) 

I don't think this is true. The target can be relative to the link. The link cares about where the file is, not vice-versa.

> and I assume it would be broken on Windows as well (I assume the link targets are not resolved relative to our CWD).

Windows uses the `relative_path` value because hard links and junctions use inodes, not paths, to define links.


Following your example, I'm working on Mac and the scenario you described produces functional links:

working with directory:
```bash
smillie@hejduk /tmp % cd root 
smillie@hejduk root % pwd
/tmp/root
smillie@hejduk root % mkdir -p x1/y1/z1
smillie@hejduk root % python -c "import os;os.symlink('x1/y1/z1', 's1')"
smillie@hejduk root % ls -l
total 0
lrwxr-xr-x  1 smillie  wheel   8 Jun 14 10:44 s1 -> x1/y1/z1
drwxr-xr-x  3 smillie  wheel  96 Jun 14 10:43 x1
smillie@hejduk root % file s1
s1: directory
```

sanity check, link should be broken now
```bash
smillie@hejduk root % rm -r x1/y1/z1 
smillie@hejduk root % ls -l         
total 0
lrwxr-xr-x  1 smillie  wheel   8 Jun 14 10:44 s1 -> x1/y1/z1
drwxr-xr-x  3 smillie  wheel  96 Jun 14 10:43 x1
smillie@hejduk root % file s1       
s1: broken symbolic link to x1/y1/z1
```

removing the now-broken link
```bash
smillie@hejduk root % rm s1
```

checking with file
```bash
smillie@hejduk root % touch x1/y1/z1
smillie@hejduk root % python -c "import os;os.symlink('x1/y1/z1', 's1')"
smillie@hejduk root % ls -l                                             
total 0
lrwxr-xr-x  1 smillie  wheel   8 Jun 14 10:45 s1 -> x1/y1/z1
drwxr-xr-x  3 smillie  wheel  96 Jun 14 10:43 x1
smillie@hejduk root % file s1                                           
s1: empty
```

Reply:
> Windows uses the relative_path value because hard links and junctions use inodes, not paths, to define links

OK I misunderstood: since there's no issue with different creation-time CWD vs. later CWD on Windows, it shouldn't be a problem.

> Following your example, I'm working on Mac and the scenario you described produces functional links:

I agree that on Mac/Linux that the correct thing happens, because we check if source_path exists relative to the link path, and then use it as is if it does).

On further examination, this also looks good for Windows: `relative_path` is correct as long as `source_path` exists relative to our CWD or to the directory containing the intended link file.